soc: esp32: Support booting from MCUboot

shubhamkulkarni97 opened the pull request:
Following changes are made to support booting Zephyr application from MCUboot:
* Removed ifdefs for IDF bootloader in the startup code and driver initialisation.
* Added mcuboot, slot-0, slot-1 and scratch partitions in device tree.
* Updated linker script to create XiP image with unsegmented image model which removes dependency on esptool to generate binary file. Generated elf file has proper LMA and VMA.
* Updated startup code to map IROM and DROM segments and setup XiP. This change is required because Espressif MCUboot port loads IRAM and DRAM regions only.
* Updated build system to build MCUboot bootloader as per the partitions provided in device tree.

**NOTE 1:** To build MCUBoot for Espressif board (ESP32), check this: 
https://github.com/zephyrproject-rtos/mcuboot/blob/main/docs/readme-espressif.md
It is not possible yet to build the porting using `west` as that porting isn't Zephyr-like.

**NOTE 2:** Booting from ESP-IDF bootloader is still the default option.

**NOTE 3:** Running `samples/hello_world` can be achieved by:
```c
west build -b esp32 samples/hello_world/ -- -DCONFIG_BOOTLOADER_MCUBOOT=y
west flash --bin-file build/zephyr/zephyr.signed.confirmed.bin
west espressif monitor
```
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/36580

sylvioalves commented:
Hi, do not merge this yet. 

NOTE 1: This PR is dependent on upstream MCUboot PR (mcu-tools/mcuboot#1032). So Zephyr's mcuboot module must be updated before merging this PR.

mbolivar-nordic reviewed (changes requested):
Sorry to NAK but I think this PR is not mergeable.

shubhamkulkarni97 committed:
module: mcuboot: Add config option to support unsigned binary generation

This commit adds a change to support running west sign command even if
the keyfile is not provided. Default value of the configuration
is set to n in order to maintain backward compatibility.

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc: esp32: Update startup code to map IROM and DROM segments

This change updates startup code to setup MMU mappings

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc: esp32: Linker script updates to boot with MCUboot

This change creates XiP image with proper LMA and VMA to boot with
MCUboot

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
runners: esp32.py: Support flashing for MCUboot and IDF bootloader

IDF bootloader requires partition table which is not a requirement for
MCUboot. Hence, esp32.py is updated to flash partition table only if it
exists.

Current implementation only allows flashing zephyr.bin. This commit adds
capability to flash binaries with other names as well
(e.g. - zephyr.signed.bin)

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
boards: esp32: Update device tree to add partitions

Adds MCUboot, primary, secondary and scratch partitions

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
soc: esp32: Remove bootloader specific ifdefs in startup code

Booting ESP32 without 2nd stage bootloader is not supported.
Hence, ifdefs in startup code are removed

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 commented on boards/xtensa/esp32/CMakeLists.txt:
```diff
@@ -60,5 +60,93 @@ if(CONFIG_BOOTLOADER_ESP_IDF)
+  add_custom_target(mcub_conf DEPENDS ${mcuboot_build_dir}/mcuboot.defconfig)
+
+  ExternalProject_Add(
+    McuBoot
```
> Since this support is for the entire esp32 architecture, I similarly think that putting it in a single board directory looks like the wrong architectural design and this should be factored out so it can be reused by all boards in the arch.

@mbolivar-nordic  I agree with your comment. I have moved bootloader building code to SoC specific CMakeLists.txt, so the change will be applicable to all ESP32 based boards. Please check if the current implementation looks fine.

sylvioalves replied:
@tejlmand @mbolivar-nordic, was the explanation above enough for you to approve this approach? If not, can you suggest add this into the TSC agenda or dev-review we can talk and move forward on this? Thanks.

tejlmand replied:
> ESP32 has a hard dependency on second stage bootloader for setting up XiP. None of the samples would work without second stage bootloader.

Might be.
But how is that different from any other sample where MCUBoot is enabled ?
If a bootloader is enabled, you must flash that bootloader before the sample will work.

Our guide clearly states that a bootloader must be flashed independently:
https://docs.zephyrproject.org/latest/guides/device_mgmt/dfu.html#bootloaders
> You need to build and flash MCUboot itself on your device

Zephyr has no dependency to the build of the bootloader in this case.
If I disable the `ExternalProject` code, then Zephyr builds just fine:
```
[140/140] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
     mcuboot_hdr:          32 B         32 B    100.00%
        metadata:          32 B         32 B    100.00%
             ROM:       74996 B    4194240 B      1.79%
     iram0_0_seg:       11460 B       128 KB      8.74%
     irom0_0_seg:        9524 B      3264 KB      0.28%
     dram0_0_seg:        1576 B     180736 B      0.87%
     dram0_1_seg:         53 KB     113840 B     47.67%
     drom0_0_seg:        1832 B    4194240 B      0.04%
    rtc_iram_seg:          0 GB         8 KB      0.00%
    rtc_slow_seg:          0 GB         4 KB      0.00%
        IDT_LIST:          0 GB         8 KB      0.00%
```

That is very different from the TF-M case.
The TF-M build produces libraries and object files that is needed by the Zephyr build system in order to link Zephyr itself:
```
arm-zephyr-eabi-gcc
  ...
  -o zephyr/zephyr_prebuilt.elf
  ...
  tfm/platform/libplatform_ns.a  tfm/app/libtfm_api_ns.a  tfm/secure_fw/s_veneers.o
  ...
```
if I try to disable `TF-M` external project on such build, then i'll get:
```
$ ninja -Cbuild
ninja: Entering directory `build'
ninja: error: 'tfm/app/libtfm_api_ns.a', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```
or
```
$ ninja -Cbuild
ninja: Entering directory `build'
ninja: error: 'tfm/secure_fw/s_veneers.o', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```

It has been discussed before at TSC that Zephyr should not do multi-image builds in it's CMake build system just for convenience of the user.
Multi image build should be kept outside Zephyr build system.
The reason TF-M is special is because the built output from TF-M is needed for the linking stage of Zephyr itself.

Looking at history I notice that Espressif has been doing this for long time starting here:
https://github.com/zephyrproject-rtos/zephyr/pull/8383

Therefore I think this should be brought up again and discussed at the TSC.
And if approved there, note down under which circumstances the Zephyr build system should allow this approach.

I'm not against introducing multi-image builds and flashing for the user-convenience, but I would like TSC approval on this with clear guidelines on when / how this should be handled. Especially because this question has be raised before.

sylvioalves replied:
Closing this thread as this has been changed and no longer make sense.

shubhamkulkarni97 commented on boards/xtensa/esp32/CMakeLists.txt:
```diff
@@ -60,5 +60,93 @@ if(CONFIG_BOOTLOADER_ESP_IDF)
+        --build-dir "${CMAKE_BINARY_DIR}"
+        -- --confirm)
+
+      set_property(GLOBAL APPEND PROPERTY extra_post_build_commands COMMAND ${west_sign})
```
@mbolivar-nordic I have added a new commit to support binary generation even if keyfile is not specified. This commit changes mcuboot.cmake.

shubhamkulkarni97 commented on scripts/west_commands/runners/esp32.py:
```diff
@@ -18,7 +18,7 @@ class Esp32BinaryRunner(ZephyrBinaryRunner):
                  app_address, baud=921600, flash_size='detect',
                  flash_freq='40m', flash_mode='dio', espidf='espidf',
-                 bootloader_bin=None, partition_table_bin=None):
+                 bootloader_bin=None, partition_table_bin=None, app_bin=None):
```
Have updated commit message for this change

mbolivar-nordic replied:
> Have updated commit message for this change

The log says:

```
IDF bootloader requires partition table which is not a requirement for
MCUboot. Hence, esp32.py is updated to flash partition table only if it
exists.

Current implementation only allows flashing zephyr.bin. This commit adds
capability to flash binaries with other names as well
(e.g. - zephyr.signed.bin)
```

But this is already supported via the --bin-file argument (see `west flash -h`) which is already available in the runner config. We don't need this.

sylvioalves replied:
@mbolivar I agree with that, however, the `cfg` structure provided in the runner does not send the `.confirmed.bin` file in `bin_file` attribute. that is the reason of adding this custom parameter in this PR.
However, we can fix this by adding the line below in the `mcuboot.cmake` file:
`zephyr_runner_file(bin ${output}.signed.confirmed.bin)` in
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/mcuboot.cmake#L125  

With this change, attribute `bin_file` will have the proper `.confirmed.bin` file. What do you think?

sylvioalves replied:
@mbolivar-nordic regardind the comment above and `mcuboot.cmake` change, sounds good?

mbolivar-nordic replied:
I'm not sure, but wouldn't that break other targets besides esp32 that tried to use this?

sylvioalves replied:
This no longer is necesary. Closing.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -5,3 +5,131 @@ zephyr_sources(
+    "CONFIG_ESP_APPLICATION_SECONDARY_START_ADDRESS=${img_1_off}\n"
+  )
+
+  get_partition_info("${esp32_dts}" "image-scratch" img_scr_off
+	  img_scr_size)
```
would the new dts CMake functions be helpful here ?
https://github.com/zephyrproject-rtos/zephyr/blob/164029b0c5bba375e2154678fcdd0cc0a2f76fbd/cmake/extensions.cmake#L2415-L2423

shubhamkulkarni97 replied:
This function is not required, existing dt APIs do the job! I have removed it from PR

tejlmand reviewed (changes requested):
I think we should treat an empty `MCUBOOT_SIGNATURE_KEY_FILE` (no key file provided) as a wish to create unsigned images for mcuboot.
Just like `imgtool` does if you don't provide `-- key <keyfile>`.

No reason for adding extra symbols and making logic more complex than necessary. 

Of course we could consider to make it very clear to the user that unsigned images are created for mcuboot if he executes `west sign` and there is no key file.
Just to be nice to those have forgotten to specify a key file by mistake.

Just like we do in NCS if default keyfile is used:
```
       ---------------------------------------------------------
        --- WARNING: Using default MCUBoot key, it should not ---
        --- be used for production.                           ---
        ---------------------------------------------------------
```
that is, in this case print something like:
```
       ------------------------------------------------
        --- No MCUBoot key provided                ---
        --- Creating unsigned MCUBoot images       ---
        ----------------------------------------------

```

d3zd3z commented on Kconfig.zephyr:
```diff
@@ -623,6 +623,14 @@ config MCUBOOT_EXTRA_IMGTOOL_ARGS
 	  you can use this option to pass extra options to imgtool.
 	  For example, you could set this to "--version 1.2".
 
+config MCUBOOT_FORCE_IMAGE_GENERATION
```
I'd really like to discourage accidentally creating unsigned images, and would prefer the build to fail if the key is not set. If someone wants to use mcuboot without signed images, they should have to explicitly request it. Perhaps another config that allows that.

shubhamkulkarni97 replied:
@d3zd3z @tejlmand So the current approach of adding a config option to generate unsigned images should work, right? Its default value is set to `n`, so it will not accidentally generate unsigned images.

If it looks good, I'll just rename `MCUBOOT_FORCE_IMAGE_GENERATION` to `MCUBOOT_UNSIGNED_IMAGE` as per @tejlmand's suggestion.

tejlmand replied:
> I'd really like to discourage accidentally creating unsigned images, and would prefer the build to fail if the key is not set.

Thanks for commenting.
Let's keep a dedicated setting for this.

d3zd3z commented on cmake/mcuboot.cmake:
```diff
@@ -20,11 +20,13 @@ function(zephyr_mcuboot_tasks)
-    # No signature key file, no signed binaries. No error, though:
-    # this is the documented behavior.
-    return()
+  if(NOT "${CONFIG_MCUBOOT_FORCE_IMAGE_GENERATION}")
+    # Check for misconfiguration.
+    if("${keyfile}" STREQUAL "")
+      # No signature key file, no signed binaries. No error, though:
+      # this is the documented behavior.
+      return()
+    endif()
```
But, please don't do this. Can it just fail?

shubhamkulkarni97 replied:
I have kept the config option so no one accidentally generates unsigned images

shubhamkulkarni97 commented:
@d3zd3z @mbolivar-nordic @tejlmand I have made required changes, please take a look

tejlmand commented on Kconfig.zephyr:
```diff
@@ -623,6 +623,12 @@ config MCUBOOT_EXTRA_IMGTOOL_ARGS
+config MCUBOOT_GENERATE_UNSIGNED_IMAGE
+	bool "Generate unsigned binary image bootable with MCUboot"
+	help
+	  The unsigned binary image will be generated even if the MCUboot signing key is
+	  not provided.
```
and if the signing key is provided, and `MCUBOOT_GENERATE_UNSIGNED_IMAGE=y` at the same time, what happens then ?

This is not clear to the reader.
As it has been decided to keep a dedicated setting for this, then the `MCUBOOT_SIGNATURE_KEY_FILE` setting should have a `depends on !MCUBOOT_GENERATE_UNSIGNED_IMAGE` to ensure a user don't specify both.

Should the

shubhamkulkarni97 replied:
Updated help statement and added dependency as suggested

shubhamkulkarni97 commented:
@d3zd3z This is PR is blocked because Espressif's MCUboot port is not added to Zephyr's module.

We were waiting for https://github.com/zephyrproject-rtos/mcuboot/pull/60 to be merged but it seems this will take some time due to other issues.

Is it possible to add Espressif specific MCUboot commits through different PR?

tejlmand reviewed (approved):
Approved.
Some minor nits to look at, but not blocking.

shubhamkulkarni97 commented:
@tejlmand Thanks for the review! I have added suggested change.

mniestroj reviewed (changes requested):
I've got an issue when trying to build:
```
CMake Error at CMakeLists.txt:151 (add_executable):
  Cannot find source file:

    /project/bootloader/mcuboot/ext/mbedtls/library/sha256.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
```

Seems that the workaround is to call:
```
cd bootloader/mcuboot/
git submodule update --init
```
which means there are some issues with mbedTLS path.

tejlmand reviewed (changes requested):
some observations on the `ExternalProject` usage.

keenanjohnson commented:
Any help needed to close out this PR? I'm so excited for this!

sylvioalves commented:
All conversation regarding mbedTLS and its cmake path is now resolved. After the sync of https://github.com/zephyrproject-rtos/mcuboot/pull/66, mbedTLS is not necessary anymore for Espressif's mcuboot porting. It means setting `MBEDTLS_DIR` is not needed at all.

carlescufi commented:
@tejlmand @mniestroj @mbolivar-nordic please re-review

cfriedt commented:
> Any help needed to close out this PR? I'm so excited for this!

Me too 🥳

sylvioalves commented:
This PR is ready when this sync https://github.com/zephyrproject-rtos/zephyr/pull/40029 gets merged.

mbolivar-nordic reviewed (dismissed):
Sorry for the long wait. I think the extra runner kwarg is not needed, though.

tejlmand commented on soc/xtensa/esp32/Kconfig.soc:
```diff
@@ -7,6 +7,10 @@ config SOC_ESP32
 	select CLOCK_CONTROL_ESP32
 	select DYNAMIC_INTERRUPTS
+	select HAS_FLASH_LOAD_OFFSET
+	select BOOTLOADER_MCUBOOT
+	select MCUBOOT_GENERATE_UNSIGNED_IMAGE
```
I don't think the SoC in use should decide the bootloader to use.
Seems like the wrong location for such knowledge to me.

Secondly, the SoCs in Zephyr might be used in other places as well, like when building MCUBoot for Zephyr which will result in:
```
warning: BOOTLOADER_MCUBOOT (defined at Kconfig.zephyr:554) has direct dependencies !MCUBOOT with value n, but is currently being y-selected by the following symbols:
 - SOC_ESP32 (defined at soc/xtensa/esp32/Kconfig.soc:4), with value y, direct dependencies <choice> (value: y), and select condition <choice> (value: y)

error: Aborting due to Kconfig warnings
```
```

sylvioalves replied:
I updated this to only enable this symbols if `BOOTLOADER_MCUBOOT` is set in SoC level. PTAL.

tejlmand commented on soc/xtensa/esp32/Kconfig.soc:
```diff
@@ -246,4 +250,25 @@ config ESP32_UNIVERSAL_MAC_ADDRESSES
 	default 2 if ESP32_UNIVERSAL_MAC_ADDRESSES_TWO
 	default 4 if ESP32_UNIVERSAL_MAC_ADDRESSES_FOUR
 
+config ESP_BOOTLOADER_SIZE
+	hex
```
What is this setting ?
Where is it used and why.

Who should set this value.
Why is it promptless.

Same goes for the settings below.

sylvioalves replied:
Not used, closing this thread.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -5,3 +5,146 @@ zephyr_sources(
+  set(mcuboot_esp_dir ${mcuboot_dir}/boot/espressif)
+  set(mcuboot_build_dir ${CMAKE_BINARY_DIR}/mcuboot)
+
+  file(COPY ${CMAKE_CURRENT_LIST_DIR}/mcuboot.defconfig DESTINATION ${mcuboot_build_dir})
+
+  dt_nodelabel(dts_partition_path NODELABEL "boot_partition")
+  dt_reg_size(mcub_size PATH ${dts_partition_path})
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_BOOTLOADER_SIZE=${mcub_size}\n")
+
+  dt_nodelabel(dts_partition_path NODELABEL "slot0_partition")
+  dt_reg_addr(img_0_off PATH ${dts_partition_path})
+  dt_reg_size(img_size PATH ${dts_partition_path})
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_APPLICATION_PRIMARY_START_ADDRESS=${img_0_off}\n"
+    "CONFIG_ESP_APPLICATION_SIZE=${img_size}\n")
+
+  dt_nodelabel(dts_partition_path NODELABEL "slot1_partition")
+  dt_reg_addr(img_1_off PATH ${dts_partition_path})
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_APPLICATION_SECONDARY_START_ADDRESS=${img_1_off}\n"
+  )
+
+  dt_nodelabel(dts_partition_path NODELABEL "scratch_partition")
+  dt_reg_addr(img_scr_off PATH ${dts_partition_path})
+  dt_reg_size(img_scr_size PATH ${dts_partition_path})
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_SCRATCH_OFFSET=${img_scr_off}\n"
+    "CONFIG_ESP_SCRATCH_SIZE=${img_scr_size}\n"
+  )
```
No explanation is given.

Care to explain why it's not simply possible to build MCUboot for Zephyr for the esp32 board, like this:
```
$ cd <workspace>/bootloader/mcuboot/boot/zephyr
$ cmake -GNinja -DBOARD=esp32 -Bbuild .
```
and then make use of the existing Zephyr DTS and Kconfig infrastructure instead of starting to copy config settings into the build folder before building MCUBoot.

As I see this, it makes it almost impossible to build MCUboot independently.

sylvioalves replied:
Not used, closing this thread.

tejlmand reviewed (changes requested):
Still not ready for merge.

tejlmand commented:
I think this multiimage approach discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/37872#discussion_r744539355
should be brought up at the next TSC.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,148 @@
+    ${CMAKE_BINARY_DIR}/mcuboot/mcuboot_esp32.elf
+    )
+
+  add_dependencies(app McuBoot)
```
Why is `app` depending on `McuBoot` ?

sylvioalves replied:
Not used, closing this thread.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,148 @@
+    McuBoot
+    SOURCE_DIR ${mcuboot_esp_dir}
+    BINARY_DIR ${mcuboot_build_dir}
+    CONFIGURE_COMMAND ""
```
Why double quotes here ?
```suggestion
    CONFIGURE_COMMAND
```

sylvioalves replied:
Build fails without this quote. Can we keep it?

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,148 @@
+    BINARY_DIR ${mcuboot_build_dir}
+    CONFIGURE_COMMAND ""
+    ${CMAKE_COMMAND} -G${CMAKE_GENERATOR}
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_CURRENT_LIST_DIR}/mcuboot-esp32.cmake
```
adding a toolchain file, which doesn't really look like toolchain file because it doesn't configure anything regarding CMAKE_C_COMPILER or similar.
https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html

It set some compile flags as cache variables, so why isn't that done as pre-load ?
`cmake -C <file>`
https://cmake.org/cmake/help/latest/manual/cmake.1.html#options

Note, even in that case, why does the `mcuboot-esp32.cmake` file belong in the Zephyr CMake build system.

It's not required in any way to build Zephyr itself.
This should be part of the TSC meeting.

sylvioalves replied:
Not used, closing this thread.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,148 @@
+
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND
+    ${PYTHON_EXECUTABLE} ${ESP_IDF_PATH}/components/esptool_py/esptool/esptool.py
```
Where do I find `esptool.py` ?

It's not here: 
https://github.com/zephyrproject-rtos/zephyr/blob/9bea59a8c4b70ebd6da5fe2cd1566eb93ebbbdcf/boards/common/esp32.board.cmake#L13

```
/usr/bin/python3.8: can't open file '/projects/github/ncs/modules/hal/espressif/components/esptool_py/esptool/esptool.py': [Errno 2] No such file or directory
ninja: build stopped: subcommand failed.
```

Note, Zephyr doesn't clone submodules.
If this is required toll, please ensure it's part of the repositories handled be `west`.

sylvioalves replied:
esptool is part of the submodules in `hal_espressif`.  
You have got to run `west espressif update` to retrieve that. It is part of the ESP32 guidelines.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,148 @@
+    ${CMAKE_BINARY_DIR}/mcuboot/mcuboot_esp32.elf
+    )
+
+  add_dependencies(app McuBoot)
```
Why is `app` depending on `McuBoot` ?

sylvioalves replied:
Not used, closing this thread.

tejlmand reviewed (changes requested):
Some more comments.

Also you need a manifest update to ensure `mbedtls` is available when testing this.
I still get:
```
CMake Error at CMakeLists.txt:151 (add_executable):
  Cannot find source file:

    /projects/github/ncs/bootloader/mcuboot/ext/mbedtls/library/sha256.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
```

when testing this PR.

beriberikix commented:
Really looking forward to seeing this merged - we have several users waiting on it! 🙏

beriberikix commented:
Just another nudge to move this along - we continue to use this in production in our fork and every day another user is asking for this to be merged!

cfriedt commented:
@shubhamkulkarni97 - it looks like there are still some changes requested by @mbolivar-nordic and @tejlmand , unless I'm mistaken.

sylvioalves commented:
@cfriedt @beriberikix Hi guys, I am working on the requested changes. Will submit soon.

sylvioalves committed:
soc: esp32: add support to mcuboot build

Select configuration when mcuboot is enabled.
At this moment, only UNSIGNED image is supported as per
Espressif's mcuboot porting.

This also updates esp32 runner to use proper bin_file name
and updates default bootloader check.

Move CMakeLists.txt content from board to soc directory
as this support is for esp32 architecture and not board
specific.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
I have updated this MR description and also make this build single-image when it comes to mcuboot support.  
@shubhamkulkarni97 @tejlmand PTAL.

sylvioalves commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -3,4 +3,139 @@
+    cmake -DCMAKE_TOOLCHAIN_FILE=${mcuboot_esp_dir}/tools/toolchain-esp32.cmake
+    -DMCUBOOT_TARGET=esp32 -DMCUBOOT_CONFIG_FILE=${mcuboot_build_dir}/mcuboot.defconfig
+    -DIDF_PATH=${ESP_IDF_PATH} -S ${mcuboot_esp_dir}
+    -GNinja &&
```
No longer necessary, closing.

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -5,3 +5,77 @@ zephyr_sources(
   loader.c
   )
+
+  if(CONFIG_BOOTLOADER_ESP_IDF)
```
minor nit.
The indentation of the `if()` is of.
```suggestion
if(CONFIG_BOOTLOADER_ESP_IDF)
```

tejlmand commented on soc/xtensa/esp32/CMakeLists.txt:
```diff
@@ -5,3 +5,77 @@ zephyr_sources(
   )
+
+  if(CONFIG_BOOTLOADER_ESP_IDF)
+  include(ExternalProject)
```
Nothing wrong on this line, but the commit message in: https://github.com/zephyrproject-rtos/zephyr/pull/37872/commits/2c6aee2cb79d0e42f87a04a673412677037efbe9

doesn't have a single line mentioning why the IDF Bootloader CMake related code is being moved from `boards/xtensa/esp32`  ` soc/xtensa/esp32/`.

Please describe this change in the commit message, as I don't see the relation to the other changes described in that message.

sylvioalves replied:
@tejlmand, It was moved so `soc` folder as per Marti's request:  
 `Since this support is for the entire esp32 architecture, I similarly think that putting it in a single board directory looks like the wrong architectural design and this should be factored out so it can be reused by all boards in the arch.`  
 I do agree that all this cmake stuff should be placed in `soc` level instead of `board` level. I'll add this into commit message as well.

tejlmand replied:
> @tejlmand, It was moved to soc folder as per Marti's request:

yes, I know why because I followed the discussion and I agree it's better to have in the SoC, my point was only to have this reason described in the commit message.

You can't expect people in the future to go back to the PR itself and read 100+ comments to figure out why something was done.

tejlmand reviewed (commented):
some minor nits.

Besides those nits, the changes looks good.

beriberikix commented:
Gentle ping - anything else required to get this merged? 🙏

sylvioalves commented:
@nashif @cfriedt PR is ready. All comments have been solved.
@mbolivar-nordic Marti, you might want to review again?

beriberikix commented:
Happy Friday! We continue to patiently await for this to be approved and merged 🙏 🙏 🙏

cfriedt commented:
> Sorry for the long wait. I think the extra runner kwarg is not needed, though.

@sylvioalves - was this comment by @mbolivar-nordic addressed? If so, I'll dismiss it.

Need to rebase. I'll merge as soon as it's green.

sylvioalves commented:
> > Sorry for the long wait. I think the extra runner kwarg is not needed, though.
> 
> @sylvioalves - was this comment by @mbolivar-nordic addressed? If so, I'll dismiss it.
> 
> Need to rebase. I'll merge as soon as it's green.

His comments were regarding the multi-image support and some code related to that. However, multi-image implementation has been removed from this PR, so it can be dismissed. Thanls @cfriedt.

maweicoltsmart commented:
@[shubhamkulkarni97](https://github.com/shubhamkulkarni97)The application booted with mcuboot runs on the esp32 soc, but when I call the function flash_img_buffered_write() to write the new firmware to the image-1 slot and restart, it doesn't work to switch to the new application, and still runs the old one program of. I don't know where is the error. The log of mcuboot is as follows:
```
[esp32] [INF] *** Booting MCUboot build v1.7.0-485-ge86f575f ***
[esp32] [INF] Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[esp32] [INF] Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[esp32] [INF] Boot source: primary slot
[esp32] [INF] Swap type: none
[esp32] [INF] Disabling RNG early entropy source...
[esp32] [INF] br_image_off = 0x10000
[esp32] [INF] ih_hdr_size = 0x20
[esp32] [INF] Loading image 0 - slot 0 from flash, area id: 1
[esp32] [INF] DRAM segment: start=0x7914, size=0xdfc, vaddr=0x3ffb0000
[esp32] [INF] IRAM segment: start=0x8710, size=0x12464, vaddr=0x40080000
[esp32] [INF] start=0x40091f74

```

sylvioalves commented:
@maweicoltsmart can you please open an issue? I'll take a look on it. Btw, how are you updating the firmware?

maweicoltsmart commented:
[sylvioalves](https://github.com/sylvioalves)Sign my application with this command python3 I use this command: python3 bootloader/mcuboot/scripts/imgtool.py sign --align 4 -v 0 -H 32 -S 0x100000 ./zephyr.bin ./app_signed.bin to sign my app. Burn the signed program and mcuboot.bin to my target board and it works fine. I signed the new version of the application in the same way and transferred it to the target board through http, and then restarted, mcuboot did not boot the new application. My other ARM-based target board also updated the firmware in this way without any problem. Is it because I did not enable secure boot when compiling mcuboot? I am using the latest code from the main branch

almir-okato commented:
Hi @maweicoltsmart, please try adding `--pad` when signing your image. 
Also, if your updated image does not confirm itself when running the first time, at the next boot it will be reverted to the old one. You can add `--confirm` if you want to avoid the revert.

One more thing, I recommend you to update your repositories to use the newest versions supported. Zephyr already signs the image when building and flashing with `west build`/`west flash` command for ESP32.

maweicoltsmart commented:
@[almir-okato](https://github.com/almir-okato) @[sylvioalves](https://github.com/sylvioalves) I compile and burn with the following commands:
```
west build -p auto -b my_esp32_board samples/myapp/
west flash
```
mcuboot & myapp works fine。
So how do I make a new version of the app firmware next? There are three bin files in the compiled build directory: zephyr.bin zephyr.signed.bin and zephyr.signed.confirmed.bin which one should I use? Which interfaces should I call to write the new version of firmware into the image-1 slot, do I need to call other interfaces after writing? The interface I am calling now is as follows:
```
flash_area_erase();
flash_img_init();
flash_img_buffered_write(）
```

Is there any problem with the above operation? My other ARM target boards do this without any problem. I don't know if there is anything special about esp32. Or can you guys give an example for reference, thanks

maweicoltsmart commented:
@[almir-okato](https://github.com/almir-okato) @[sylvioalves](https://github.com/sylvioalves) I compile and burn with the following commands:
```
west build -p auto -b my_esp32_board samples/myapp/
west flash
```
mcuboot & myapp works fine。
So how do I make a new version of the app firmware next? There are three bin files in the compiled build directory: zephyr.bin zephyr.signed.bin and zephyr.signed.confirmed.bin which one should I use? Which interfaces should I call to write the new version of firmware into the image-1 slot, do I need to call other interfaces after writing? The interface I am calling now is as follows:
```
flash_area_erase();
flash_img_init();
flash_img_buffered_write(）
```

Is there any problem with the above operation? My other ARM target boards do this without any problem. I don't know if there is anything special about esp32. Or can you guys give an example for reference, thanks

maweicoltsmart commented:
My guess is that the mcuboot I am using doesn't check if there is any firmware available in the mage-1 slot, can you please send your compiled mcuboot.bin to my email? Thanks! My email address is: ***@***.***

--------------原始邮件--------------
发件人："almir-okato ***@***.***&gt;;
发送时间：2022年5月11日(星期三) 晚上9:05
收件人："zephyrproject-rtos/zephyr" ***@***.***&gt;;
抄送："maweicoltsmart ***@***.***&gt;;"Mention ***@***.***&gt;;
主题：Re: [zephyrproject-rtos/zephyr] soc: esp32: Support booting from MCUboot (#37872)
-----------------------------------

 

 
@maweicoltsmart use the zephyr.signed.confirmed.bin binary, unless you want your app to manually confirm itself.
 
Regarding the the flash writing, please check if you properly configured Zephyr's Flash subsystem and DFU subsystem. You can take a look on Hawkbit example for image update on zephyr/subsys/mgmt/hawkbit.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

maweicoltsmart commented:
@[almir-okato](https://github.com/almir-okato)My guess is that the mcuboot I am using doesn't check if there is any firmware available in the mage-1 slot, can you please send your compiled mcuboot.bin to my email? Thanks! 

Here is the log I reboot after writing the new firmware to the image-1 slot:
```
I: Download progress: 1048576 Bytes; 1024 KiB; 1 MiB
I: Finished downloading. Reboot ...
[esp32] [INF] Enabling RNG early entropy source...
[esp32] [INF] *** Booting MCUboot build v1.7.0-485-ge86f575f ***
[esp32] [INF] Primary image: magic=bad, swap_type=0x1, copy_done=0x3, image_ok=0x3
[esp32] [INF] Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[esp32] [INF] Boot source: none
[esp32] [INF] Swap type: none
[esp32] [INF] Disabling RNG early entropy source...
[esp32] [INF] br_image_off = 0x10000
[esp32] [INF] ih_hdr_size = 0x20
[esp32] [INF] Loading image 0 - slot 0 from flash, area id: 1
[esp32] [INF] DRAM segment: start=0x79ec, size=0xdfc, vaddr=0x3ffb0000
[esp32] [INF] IRAM segment: start=0x87e8, size=0x12464, vaddr=0x40080000
[esp32] [INF] start=0x40091f74
D: Generic GSM modem (0x3ffb6940)
D: iface->read 0x400e408c iface->write 0x400e40e0
D: starting
I: gsm_ppp_start, 1442
D: Starting modem 0x3ffb6940 configuration
I: Setting Modem Pins
*** Booting Zephyr OS build zephyr-v3.0.0-3556-g81cc3626aca7  ***
I: Power on
```

almir-okato commented:
@maweicoltsmart please paste here the MCUboot `bootloader.conf` file that you are using, it should be located in `mcuboot/boot/espressif/port/esp32/` or `mcuboot/boot/espressif/`, and also check if the `MCUBOOT_GENERATE_CONFIRMED_IMAGE` option is enabled in Zephyr.

The checking for firmware available in the image-1 slot is part of MCUboot process and it depends on the image TLVs (metadata used by MCUboot) appended when using `imgtool` to sign.

Test the update using this app (it's a HelloWorld) with your bootloader:
[zephyr.signed.confirmed.zip](https://github.com/zephyrproject-rtos/zephyr/files/8671188/zephyr.signed.confirmed.zip)

If it didn't work, try with this bootloader:
[mcuboot_esp32.zip](https://github.com/zephyrproject-rtos/zephyr/files/8671038/mcuboot_esp32.zip)

maweicoltsmart commented:
@[sylvioalves](https://github.com/sylvioalves)I tried to compile your simple_http_ota/ and got an error

```
colt@colt-virtualbox:~/zephyrproject/code/zephyr$ west build -p auto -b esp32 samples/boards/esp32/simple_http_ota/
[1/11] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/colt/zephyrproject/code/zephyr/build/zephyr/zephyr_pre0.map 
: && ccache /home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/home/colt/zephyrproject/code/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/subsys/dfu/boot/libsubsys__dfu__boot.a  zephyr/subsys/net/libsubsys__net.a  zephyr/subsys/net/l2/ethernet/libsubsys__net__l2__ethernet.a  zephyr/subsys/net/ip/libsubsys__net__ip.a  zephyr/subsys/net/lib/http/libsubsys__net__lib__http.a  zephyr/subsys/random/libsubsys__random.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/i2c/libdrivers__i2c.a  zephyr/drivers/wifi/libdrivers__wifi.a  zephyr/drivers/hwinfo/libdrivers__hwinfo.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  modules/mbedtls/libmodules__mbedtls.a  modules/mcuboot/libmcuboot_util.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0"  -L/home/colt/zephyrproject/code/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/src/linker/esp32.rom.alias.ld  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.ld  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.api.ld  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.libgcc.ld  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.redefined.ld  -T/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp32/ld/esp32.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  -lnet80211  -lcore  -lpp  -lcoexist  -lphy  -lrtc  -lgcc  -L/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp32  -L/home/colt/zephyrproject/code/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/lib/esp32  -Wl,--defsym,phy_printf=printf  -Wl,--defsym,net80211_printf=printf  -Wl,--defsym,coexist_printf=printf && cd /home/colt/zephyrproject/code/zephyr/build/zephyr && /usr/bin/cmake -E echo
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lnet80211
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lcore
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lpp
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lcoexist
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lphy
/home/colt/.espressif/tools/zephyr/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.bfd: cannot find -lrtc
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/colt/zephyrproject/code/zephyr/build
```

maweicoltsmart commented:
@[almir-okato](https://github.com/almir-okato)
@[sylvioalves](https://github.com/sylvioalves)
I found that when I downloaded the firmware via http and wrote it to the image-1 slot, there was an error when calling boot_request_upgrade(), the return value was -14

sylvioalves commented:
@maweicoltsmart Can you confirm which Zephyr revision you are working with?  
Also, confirm that  `west update`  and `west espressif update` is executed.

ycsin commented:
> @ycsin, Do you think it is worth having this sample code above in Zephyr? Not sure there is already something very straightforward like this.

We have something that downloads file using http but that's it. To me I don't mind having this upstreamed, I think having more perspective is always good for others to refer to and learn. I think you can try to clean it up abit (i.e.: `http_client` #43236 has been updated to calc `body_frag_len` #43237) and do a PR then see what @rlubos think and others think.

ashegheparyshun commented:
> @maweicoltsmart, I have uploaded a simple HTTP OTA sample code that uses MCUBoot and a python3 HTTP server. https://github.com/sylvioalves/zephyr/tree/feature/http_ota_sample
> 
> It just works. Can you take a look and see if that helps you in any way?
> 
> @ycsin, Do you think it is worth having this sample code above in Zephyr? Not sure there is already something very straightforward like this.
> 
> Thanks.

I think it worth. It helped me a lot.
TNX.
